libraries/AP_Proximity/AP_Proximity_ARS408_CAN.cpp

```diff
@@ -0,0 +1,388 @@
+    send_speed_message();
+}
+
+void AP_Proximity_ARS408_CAN::send_speed_message()
+{
+    float speed = 0.0f;
+    if (!get_forward_speed(speed)) {
+        // we expect caller will not try to control heading using rate control without a valid speed estimate
+        // on failure to get speed we do not attempt to steer
+        return;
+    }
+
+    // Set speed
+    uint32_t radar_speed = static_cast<uint32_t>(fabs(speed) / 0.02);
+    speed_information speed_information_msg;
+    speed_information_msg.data.RadarDevice_Speed1 = static_cast<uint64_t>(radar_speed >> 8);
+    speed_information_msg.data.RadarDevice_Speed2 = static_cast<uint64_t>(radar_speed & 255);
+    speed_information_msg.data.RadarDevice_SpeedDirection = RadarDevice_SpeedDirection::STANDSTILL;
+    if(speed > 0.2){speed_information_msg.data.RadarDevice_SpeedDirection  = RadarDevice_SpeedDirection::FORWARD;}
+    if(speed < -0.2){speed_information_msg.data.RadarDevice_SpeedDirection = RadarDevice_SpeedDirection::BACKWARD;}
+
+    // construct AP_HAL::CANframe
+    AP_HAL::CANFrame speed_frame;
+    speed_frame = {(0x300 & AP_HAL::CANFrame::MaskStdID), speed_information_msg.raw_data, sizeof(speed_information_msg.raw_data)};
+    write_frame(speed_frame, AP_HAL::native_micros64() + 1000);
+
+}
+
+// get forward speed in m/s (earth-frame horizontal velocity but only along vehicle x-axis).  returns true on success
+bool AP_Proximity_ARS408_CAN::get_forward_speed(float &speed) const
+{
+    Vector3f velocity;
+    const AP_AHRS &_ahrs = AP::ahrs();
+    if (!_ahrs.get_velocity_NED(velocity)) {
+        // use less accurate GPS, assuming entire length is along forward/back axis of vehicle
+        if (AP::gps().status() >= AP_GPS::GPS_OK_FIX_3D) {
+            if (abs(wrap_180_cd(_ahrs.yaw_sensor - AP::gps().ground_course_cd())) <= 9000) {
+                speed = AP::gps().ground_speed();
+            } else {
+                speed = -AP::gps().ground_speed();
+            }
+            return true;
+        } else {
+            return false;
+        }
+    }
+    // calculate forward speed velocity into body frame
+    speed = velocity.x*_ahrs.cos_yaw() + velocity.y*_ahrs.sin_yaw();
+    return true;
+}
+
+
+void AP_Proximity_ARS408_CAN::send_yaw_rate_message()
+{
+    float yaw_rate = -degrees(AP::ahrs().get_yaw_rate_earth());
+    uint32_t radar_yaw_rate = static_cast<uint32_t>((yaw_rate + 327.68) / 0.01);
+
+    yaw_rate_information yaw_rate_raw;
+    yaw_rate_raw.data.RadarDevice_YawRate1 = static_cast<uint64_t>(radar_yaw_rate >> 8);
+    yaw_rate_raw.data.RadarDevice_YawRate2 = static_cast<uint64_t>(radar_yaw_rate & 255);
+
+    // construct CANFrame
+    AP_HAL::CANFrame yaw_rate_frame;
+    yaw_rate_frame = {(0x301 & AP_HAL::CANFrame::MaskStdID), yaw_rate_raw.raw_data, sizeof(yaw_rate_raw.raw_data)};
+    write_frame(yaw_rate_frame, AP_HAL::native_micros64() + 1000);
+}
+
+
+bool AP_Proximity_ARS408_CAN::set_max_distance(uint64_t distance, bool valid) {
```

Comment:
Should we really take a distance in metres in as a uint64_t ?  Perhaps take it as a uint16_t instead?